From bf0a751e2780f9bf8ee1c69397e0809f4df3d988 Mon Sep 17 00:00:00 2001
From: Jian Liang <jian.l@ssi.samsung.com>
Date: Wed, 16 Dec 2020 09:47:02 -0800
Subject: [PATCH 16/78] * blobfs bug fix: __read_done should return bserror

---
 lib/blobfs/blobfs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/blobfs/blobfs.c b/lib/blobfs/blobfs.c
index 536f374..e427784 100644
--- a/lib/blobfs/blobfs.c
+++ b/lib/blobfs/blobfs.c
@@ -1859,7 +1859,7 @@ __read_done(void *ctx, int bserrno)
 	        if(args->fn.file_op == __rw_from_file_done){
 			_copy_buf_to_iovs(args->iovs, args->iovcnt, buf, args->op.rw.length);
 		}
-		__rw_done(req, 0);
+		__rw_done(req, bserrno);
 	} else {
 		_copy_iovs_to_buf(buf, args->op.rw.length, args->iovs, args->iovcnt);
 		spdk_blob_io_write(args->file->blob, args->op.rw.channel,
-- 
1.8.3.1

