From 6f065877e551e21c9fe2b25bde79234c5796c99b Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Thu, 4 Mar 2021 12:49:08 -0800
Subject: [PATCH 33/78] Reap request from qos only when enabled

---
 lib/nvmf/rdma.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/nvmf/rdma.c b/lib/nvmf/rdma.c
index 0b1df92..5f4ccb5 100644
--- a/lib/nvmf/rdma.c
+++ b/lib/nvmf/rdma.c
@@ -735,6 +735,10 @@ static inline void dfly_poller_rdma_qos_sched(struct spdk_nvmf_rdma_transport *r
 	struct dfly_request *w[64];
 	struct spdk_nvmf_rdma_request *nvmf_req;
 
+	if(!g_dragonfly->df_qos_enable) {
+		return;
+	}
+
 	TAILQ_FOREACH(rqpair, &rpoller->qpairs, link) {
 		i = 0;
 
-- 
1.8.3.1

