From 1d871f227d40ccdb3c70767d25dda50b5d1a35ab Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Tue, 23 Mar 2021 14:24:03 -0700
Subject: [PATCH 39/75] Trim disable update

---
 lib/blob/blobstore.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/blob/blobstore.c b/lib/blob/blobstore.c
index 6141eb3..5af1561 100644
--- a/lib/blob/blobstore.c
+++ b/lib/blob/blobstore.c
@@ -1544,6 +1544,9 @@ bs_batch_clear_dev(struct spdk_blob_persist_ctx *ctx, spdk_bs_batch_t *batch, ui
 {
 	switch (ctx->blob->clear_method) {
 	case BLOB_CLEAR_WITH_DEFAULT:
+		if(g_bs_trim_enable == false) {
+			break;
+		}
 	case BLOB_CLEAR_WITH_UNMAP:
 		bs_batch_unmap_dev(batch, lba, lba_count);
 		break;
-- 
1.8.3.1

