From f13aa911b8c6ad50889cbc6cddc0017f5fa6fea4 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Wed, 31 Mar 2021 13:23:06 -0700
Subject: [PATCH 42/49] Modify INI deprecated log to warning

---
 lib/conf/conf.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/conf/conf.c b/lib/conf/conf.c
index 287e157..744e024 100644
--- a/lib/conf/conf.c
+++ b/lib/conf/conf.c
@@ -623,7 +623,7 @@ spdk_conf_read(struct spdk_conf *cp, const char *file)
 	if (file == NULL || file[0] == '\0') {
 		return -1;
 	}
-	SPDK_ERRLOG("INI configuration has been deprecated and will be removed in a future release. Please switch to JSON-RPC.\n");
+	SPDK_WARNLOG("INI configuration has been deprecated and will be removed in a future release. Please switch to JSON-RPC.\n");
 
 	fp = fopen(file, "r");
 	if (fp == NULL) {
-- 
1.8.3.1

