From 90d3e7af3561ea4cd94c6ecd4e357cfa12aba557 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Tue, 12 Oct 2021 13:56:33 -0700
Subject: [PATCH 50/74] Config option for DSS open source release

---
 CONFIG         | 3 +++
 mk/spdk.app.mk | 6 +++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/CONFIG b/CONFIG
index d947909e9..c8fd1a8ba 100644
--- a/CONFIG
+++ b/CONFIG
@@ -170,3 +170,6 @@ CONFIG_RAID5=n
 
 # Build with IDXD support
 CONFIG_IDXD=n
+
+#Build for DSS Open Source Release
+CONFIG_DSS_OSS=n
diff --git a/mk/spdk.app.mk b/mk/spdk.app.mk
index 559396b56..80569b744 100644
--- a/mk/spdk.app.mk
+++ b/mk/spdk.app.mk
@@ -48,7 +48,11 @@ endif
 
 LIBS += $(SPDK_LIB_LINKER_ARGS)
 
-SYS_LIBS += -ljemalloc -lsnappy -lz -lm -ldl -L$(SPDK_ROOT_DIR)/../../nkv-target/lib -L$(SPDK_ROOT_DIR)/../../ -loss -ldss_hsl -ldss_lat -ljudyL -ldssd -lstdc++  -L$(SPDK_ROOT_DIR)/../rocksdb -lrocksdb
+SYS_LIBS += -ljemalloc -lsnappy -lz -lm -ldl -L$(SPDK_ROOT_DIR)/../../nkv-target/lib -L$(SPDK_ROOT_DIR)/../../ -loss -ldssd -lstdc++  -L$(SPDK_ROOT_DIR)/../rocksdb -lrocksdb
+
+ifeq ($(CONFIG_DSS_OSS),n)
+SYS_LIBS += -ldss_hsl -ldss_lat -ljudyL
+endif
 
 CLEAN_FILES = $(APP)
 
-- 
2.24.3

