From 6e19d9762623b6207b1ebad5865cfa65c7154d47 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Thu, 28 Oct 2021 10:32:07 -0700
Subject: [PATCH 51/68] Update blobfs logging to include file pointer

---
 lib/blobfs/blobfs.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/lib/blobfs/blobfs.c b/lib/blobfs/blobfs.c
index ec115abb2..9854c1320 100644
--- a/lib/blobfs/blobfs.c
+++ b/lib/blobfs/blobfs.c
@@ -49,10 +49,10 @@
 #include "dragonfly.h"
 
 #define BLOBFS_TRACE(file, str, args...) \
-	SPDK_DEBUGLOG(SPDK_LOG_BLOBFS, "file=%s " str, file->name, ##args)
+	SPDK_DEBUGLOG(SPDK_LOG_BLOBFS, "fp=%p file=%s " str, file, file->name, ##args)
 
 #define BLOBFS_TRACE_RW(file, str, args...) \
-	SPDK_DEBUGLOG(SPDK_LOG_BLOBFS_RW, "file=%s " str, file->name, ##args)
+	SPDK_DEBUGLOG(SPDK_LOG_BLOBFS_RW, "fp=%p file=%s " str, file, file->name, ##args)
 
 #define BLOBFS_DEFAULT_CACHE_SIZE (4ULL * 1024 * 1024 * 1024)
 #define SPDK_BLOBFS_DEFAULT_OPTS_CLUSTER_SZ (1024 * 1024)
@@ -2301,8 +2301,7 @@ cache_insert_buffer(struct spdk_file *file, uint64_t offset)
 			break;
 		}
 		if (count++ == 100) {
-			SPDK_ERRLOG("Could not allocate cache buffer for file=%p on offset=%jx\n",
-				    file, offset);
+			BLOBFS_TRACE(file, "Could not allocate cache buffer on offset=%jx\n", offset);
 			free(buf);
 			return NULL;
 		}
@@ -2334,7 +2333,7 @@ cache_append_buffer(struct spdk_file *file)
 
 	last = cache_insert_buffer(file, file->append_pos);
 	if (last == NULL) {
-		SPDK_DEBUGLOG(SPDK_LOG_BLOBFS, "cache_insert_buffer failed\n");
+		BLOBFS_TRACE(file, "cache_insert_buffer failed\n");
 		return NULL;
 	}
 
-- 
2.24.3

