From 6014eff5f18fcff51ad869f35e7cb9bb0cb080e9 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Thu, 23 Dec 2021 15:48:27 -0800
Subject: [PATCH 58/70] Test spdk patch

---
 lib/nvmf/rdma.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/nvmf/rdma.c b/lib/nvmf/rdma.c
index f92f09682..461e51ee3 100644
--- a/lib/nvmf/rdma.c
+++ b/lib/nvmf/rdma.c
@@ -31,6 +31,9 @@
  *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+//Test
+//
+//
 #include "spdk/stdinc.h"
 
 #include "spdk/config.h"
-- 
2.24.3

