From eacebd43b09f6dec168235f0b9989d63f60b71f0 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Tue, 30 Aug 2022 11:53:09 -0700
Subject: [PATCH 67/67] Return failure for rdd handle not found

---
 lib/nvmf/rdma.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/lib/nvmf/rdma.c b/lib/nvmf/rdma.c
index dfed83a99..d84532523 100644
--- a/lib/nvmf/rdma.c
+++ b/lib/nvmf/rdma.c
@@ -1589,6 +1589,12 @@ request_transfer_in(struct spdk_nvmf_request *req)
 	return 0;
 }
 
+/**
+ * data_posted
+ *   1  == data posted
+ *   0  == No data posted
+ *   -1 == Error posting data
+ */
 static int
 request_transfer_out(struct spdk_nvmf_request *req, int *data_posted)
 {
@@ -1658,8 +1664,8 @@ request_transfer_out(struct spdk_nvmf_request *req, int *data_posted)
 				return 0;//Success skip nvmf data transfer
 			}  else {
 				//IO failed
-				//TODO: setup error code and complete
-				assert(0);
+                *data_posted = -1;
+                return 0;
 			}
 
 		} else {
@@ -2901,8 +2907,14 @@ nvmf_rdma_request_process(struct spdk_nvmf_rdma_transport *rtransport,
 			if (rc) {
 				rdma_req->state = RDMA_REQUEST_STATE_COMPLETED;
 			} else {
-				rdma_req->state = data_posted ? RDMA_REQUEST_STATE_TRANSFERRING_CONTROLLER_TO_HOST :
+                if(data_posted == -1) {
+                    //Set error and state to completed
+				    rsp->status.sc = SPDK_NVME_SC_INTERNAL_DEVICE_ERROR;
+				    rdma_req->state = RDMA_REQUEST_STATE_READY_TO_COMPLETE;
+                } else {
+				    rdma_req->state = data_posted ? RDMA_REQUEST_STATE_TRANSFERRING_CONTROLLER_TO_HOST :
 						  RDMA_REQUEST_STATE_COMPLETING;
+                }
 			}
 			break;
 		case RDMA_REQUEST_STATE_TRANSFERRING_CONTROLLER_TO_HOST:
-- 
2.24.3

