From 6e0d6cc52301a1458eb7d306ee4f072f60a84a3f Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Wed, 19 Apr 2023 16:21:04 -0700
Subject: [PATCH 74/78] Make rocksdb linking optional

---
 CONFIG         | 3 +++
 mk/spdk.app.mk | 6 +++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/CONFIG b/CONFIG
index c8fd1a8..8346a42 100644
--- a/CONFIG
+++ b/CONFIG
@@ -173,3 +173,6 @@ CONFIG_IDXD=n
 
 #Build for DSS Open Source Release
 CONFIG_DSS_OSS=n
+
+#Build with Rocksdb KV support
+CONFIG_DSS_ROCKSDB_KV=n
diff --git a/mk/spdk.app.mk b/mk/spdk.app.mk
index 80569b7..185dcd7 100644
--- a/mk/spdk.app.mk
+++ b/mk/spdk.app.mk
@@ -48,7 +48,11 @@ endif
 
 LIBS += $(SPDK_LIB_LINKER_ARGS)
 
-SYS_LIBS += -ljemalloc -lsnappy -lz -lm -ldl -L$(SPDK_ROOT_DIR)/../../nkv-target/lib -L$(SPDK_ROOT_DIR)/../../ -loss -ldssd -lstdc++  -L$(SPDK_ROOT_DIR)/../rocksdb -lrocksdb
+SYS_LIBS += -ljemalloc -lsnappy -lz -lm -ldl -L$(SPDK_ROOT_DIR)/../../nkv-target/lib -L$(SPDK_ROOT_DIR)/../../ -loss -ldssd -lstdc++
+
+ifeq ($(CONFIG_DSS_ROCKSDB_KV),y)
+SYS_LIBS += -L$(SPDK_ROOT_DIR)/../rocksdb -lrocksdb
+endif
 
 ifeq ($(CONFIG_DSS_OSS),n)
 SYS_LIBS += -ldss_hsl -ldss_lat -ljudyL
-- 
1.8.3.1

