From b6192235eabfe6faac7dd5dfa2827cf03634e68b Mon Sep 17 00:00:00 2001
From: Jian Liang <jian.l@ssi.samsung.com>
Date: Wed, 23 Dec 2020 23:20:51 -0800
Subject: [PATCH 19/26] * rdb options: wal, sync, auto_compaction

---
 db/dss_kv2blk_c.cc | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/db/dss_kv2blk_c.cc b/db/dss_kv2blk_c.cc
index f99dccd..f10dc39 100644
--- a/db/dss_kv2blk_c.cc
+++ b/db/dss_kv2blk_c.cc
@@ -195,7 +195,7 @@ struct RandomReadContext : public async::AsyncStatusCapture {
 
 void *_dss_rocksdb_open(void *arg)
 {
-	struct dfly_io_device_s *dss_dev = (struct dfly_io_device_s *)arg;
+    struct dfly_io_device_s *dss_dev = (struct dfly_io_device_s *)arg;
 
     rocksdb::DB* db;
     rocksdb::Options options;
@@ -236,7 +236,7 @@ void *_dss_rocksdb_open(void *arg)
     options.num_levels = 2;
     options.compression = rocksdb::kNoCompression;
     options.target_file_size_base = 1073741824; // 1GB
-    options.disable_auto_compactions = true;
+    options.disable_auto_compactions = !g_dragonfly->rdb_auto_compaction_enable;
     options.min_write_buffer_number_to_merge = 4;
     rocksdb::Status s = rocksdb::DB::Open(options, dbname, &db);
     if (!s.ok()) {
@@ -293,7 +293,6 @@ void *_dss_rocksdb_compaction(void *ctx)
 	}
 
 	pthread_mutex_unlock(&dss_dev->rdb_handle->rdb_lock);
-
 	printf("Compaction started for device %s\n", dss_dev->rdb_handle->dev_name);
 	//Syncronous compaction (Blocking)
 	s = db->CompactRange(rocksdb::CompactRangeOptions(), nullptr, nullptr);
@@ -395,8 +394,8 @@ int dss_rocksdb_put(void *ctx, void *key, uint32_t klen, void *value, uint32_t v
 
 		rocksdb::Status s;
 		rocksdb::WriteOptions write_options = rocksdb::WriteOptions();
-		write_options.sync = false;
-		write_options.disableWAL = false;
+		write_options.sync = g_dragonfly->rdb_sync_enable;
+		write_options.disableWAL = !g_dragonfly->rdb_wal_enable;
 		//printf("dss_rocksdb_put key: %s\n", (char *)key);
 		//printf("dss_rocksdb_put with wal=%d\n", !write_options.disableWAL);
 		s = db->Put(write_options, rocksdb::Slice((char *)key, klen), rocksdb::Slice((char *)value, vlen));
-- 
1.8.3.1

