From 6b8b0763fa96ccb1db65716d9411bd98db8a36b2 Mon Sep 17 00:00:00 2001
From: Benixon Arul dhas <benixon.a@samsung.com>
Date: Thu, 7 Jan 2021 13:50:34 -0800
Subject: [PATCH 23/26] Remove error log for key not found

---
 db/dss_kv2blk_c.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/db/dss_kv2blk_c.cc b/db/dss_kv2blk_c.cc
index d61aeea..42d7bc5 100644
--- a/db/dss_kv2blk_c.cc
+++ b/db/dss_kv2blk_c.cc
@@ -124,9 +124,10 @@ struct RandomReadContext : public async::AsyncStatusCapture {
 
 	  //printf("Async read completed\n");
       if (status.IsNotFound()) {
-        fprintf(stderr, "key=%s no found\n", key_.ToString(true).c_str());
+        //fprintf(stderr, "key=%s no found\n", key_.ToString(true).c_str());
 			response->status.sc = SPDK_NVME_SC_KV_KEY_NOT_EXIST;
 			response->status.sct = SPDK_NVME_SCT_KV_CMD;
+			response->status.dnr = 1;//Do not retry
       }else if(status.ok()) {
 		response->status.sc = SPDK_NVME_SC_SUCCESS;
 		response->status.sct = SPDK_NVME_SCT_GENERIC;
-- 
1.8.3.1

